tests/unit/test_dataset_records.py

```diff
@@ -72,18 +72,64 @@ def create_callback(request, snoop, status):
             partial(create_callback, snoop=snoop, status=200),
         )
 
-        self.assertRaises(HTTPError, lambda: dataset.update_records(updates))
+        self.assertRaises(HTTPError, lambda: dataset._update_records(updates))
         self.assertEqual(snoop["payload"], TestDatasetRecords.stringify(updates, False))
 
-        response = dataset.update_records(updates, ignore_nan=True)
+        response = dataset._update_records(updates, ignore_nan=True)
         self.assertEqual(response, self._response_json)
         self.assertEqual(snoop["payload"], TestDatasetRecords.stringify(updates, True))
 
+    @responses.activate
+    def test_upsert(self):
+        def create_callback(request, snoop):
+            snoop["payload"] = list(request.body)
+            return 200, {}, simplejson.dumps(self._response_json)
+
+        responses.add(responses.GET, self._dataset_url, json={})
+        dataset = self.unify.datasets.by_resource_id(self._dataset_id)
+
+        records_url = f"{self._dataset_url}:updateRecords"
+        updates = TestDatasetRecords.records_to_updates(self._records_json)
+        snoop = {}
+        responses.add_callback(
+            responses.POST, records_url, partial(create_callback, snoop=snoop)
+        )
+
+        response = dataset.upsert_records(self._records_json, "attribute1")
+        self.assertEqual(response, self._response_json)
+        self.assertEqual(snoop["payload"], TestDatasetRecords.stringify(updates, False))
+
+    @responses.activate
+    def test_delete(self):
+        def create_callback(request, snoop):
+            snoop["payload"] = list(request.body)
+            return 200, {}, simplejson.dumps(self._response_json)
+
+        responses.add(responses.GET, self._dataset_url, json={})
+        dataset = self.unify.datasets.by_resource_id(self._dataset_id)
+
+        records_url = f"{self._dataset_url}:updateRecords"
+        deletes = TestDatasetRecords.records_to_deletes(self._records_json)
+        snoop = {}
+        responses.add_callback(
+            responses.POST, records_url, partial(create_callback, snoop=snoop)
+        )
+
+        response = dataset.delete_records(self._records_json, "attribute1")
+        self.assertEqual(response, self._response_json)
+        self.assertEqual(snoop["payload"], TestDatasetRecords.stringify(deletes, False))
+
+    @staticmethod
+    def records_to_deletes(records):
+        return [
+            {"action": "DELETE", "recordId": i + 1} for i, record in enumerate(records)
```

Comment:
if you want enumeration to start at `1` instead of `0`, you can do:

```suggestion
            {"action": "DELETE", "recordId": i + 1} for i, record in enumerate(records, start=1)
```

Reply:
There are really so many python features I know nothing about